File: Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/elm/execution/UpperEvaluator.java
package org.opencds.cqf.cql.engine.elm.execution;

import org.opencds.cqf.cql.engine.exception.InvalidOperatorArgument;
import org.opencds.cqf.cql.engine.execution.Context;

/*
Upper(argument String) String

The Upper operator returns the upper case of its argument.
If the argument is null, the result is null.
*/

public class UpperEvaluator extends org.cqframework.cql.elm.execution.Upper {

    public static Object upper(Object operand) {
        if (operand == null) {
            return null;
        }

        if (operand instanceof String) {
            return ((String) operand).toUpperCase();
        }

        throw new InvalidOperatorArgument(
                "Upper(String)",
                String.format("Upper(%s)", operand.getClass().getName())
        );
    }

    @Override
    protected Object internalEvaluate(Context context) {
        Object operand = getOperand().evaluate(context);
        return upper(operand);
    }
}

File: Src/java/engine/src/main/kotlin/org/opencds/cqf/cql/engine/elm/executing/UpperEvaluator.kt
package org.opencds.cqf.cql.engine.elm.executing

import java.util.Locale
import org.opencds.cqf.cql.engine.exception.InvalidOperatorArgument

/*
Upper(argument String) String

The Upper operator returns the upper case of its argument.
If the argument is null, the result is null.
*/
object UpperEvaluator {
    @JvmStatic
    fun upper(operand: Any?): Any? {
        if (operand == null) {
            return null
        }

        if (operand is String) {
            return operand.uppercase(Locale.getDefault())
        }

        throw InvalidOperatorArgument("Upper(String)", "Upper(${operand.javaClass.name})")
    }
}
